Management of names of archived streams #26661

sanchi-t opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #26661 <!-- Issue link, or clear description.-->

Also please look at #30315.

This PR:

- Does not rename streams during deactivating them.
- Add migration to rename previously archived streams back to their original name. Also renames handful of streams pre 1b6f68bb59dc85125ff21fbd3f98a88f39c0730d.
- Add `archived` property in `RawStreamDict`, `SubscriptionStreamDict`, `APIStreamDict` and `APISubscriptionDict`.
- Add `include_archived` parameter for `[/api/get-streams](https://chat.zulip.org/api/get-streams)` and `archived_streams` as `client capability` for `[/api/register-queue](https://chat.zulip.org/api/register-queue)`.
- Displays archived stream name and messages with `(archived)` postfix to the user.
- Prevents users from sending message in an archived stream.
- Add support to rename archived streams while creating stream with the same name.

Consider following things regarding the commits while reviewing the PR.

## **Commit:** 3b9dcb51bc843dbd2a301724c8dde264f422352a
This is a prep commit that fixes the function calls in `server_events_dispatch.js`.

## **Commit:** 45ede22f9834761604e735c95afa670b687c93c9
`(Archived)` postfix is added to archived stream names by checking for archived property of the streams.
The `web/templates/message_view_header.hbs` code looks unclean and cluttered, I couldn't make it more cleaner. Also I am not aware about any other method that will eliminate nesting if condition in `web/templates/navbar_icon_and_title.hbs`.

## **Commit:** adf99b9b01f47b5e80832a47c5905d957aa57b84

A significant portion of the code for the modal window in `web/src/stream_create.js` is similar to the existing code in `web/src/stream_edit.js`. I haven't identified a method to efficiently reuse that code.




<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

## **UI Changes:**

Since Testing Stream (an active stream) already exists, the migration did not renamed the archived stream with same old name:
![image](https://github.com/zulip/zulip/assets/98596642/c1edf15a-ffa4-420f-8e5a-482d759e0de5)
![image](https://github.com/zulip/zulip/assets/98596642/55010ed8-1687-4fa2-ae87-8aa2d70e20d1)

When user doesn't has the permission to rename the archived stream:
![{CB2240E9-2266-4C24-A29B-4DD2AB663B21}](https://github.com/user-attachments/assets/fa077e5e-81c9-4e6f-941c-ae9ab8d043b4)

When user cannot view the archived stream:
![{AA807129-26A6-4012-8D39-2F7A623E0A99}](https://github.com/user-attachments/assets/fb58e787-262e-4165-bdb1-3faa7ecf62c7)

When user has the permission to rename the archived stream:
![{2A4F8BEF-894A-44E2-B324-223B9ED2BD7E}](https://github.com/user-attachments/assets/a61e1183-8d15-4c50-990a-032669383aa9)
![{EE1A5744-B552-439A-A48C-5EA0C6DB9A0B}](https://github.com/user-attachments/assets/c75d686b-8515-4a25-9472-a4e65a18d46b)


Archiving notice:
![{BF469E14-90AF-48C5-AA53-8BCBAC652643}](https://github.com/user-attachments/assets/515b6c22-58d5-42e8-8841-adebfaf7ff06)


Compose Box banner:
![{C768B5FE-6CE7-4A59-99CB-2E5EE8600EF8}](https://github.com/user-attachments/assets/ca835f0e-57d8-4db6-89e9-76ff64768a4e)

Channel archived message:
![{3AC3AA3F-7EC7-4C03-803E-AF27251B7FC1}](https://github.com/user-attachments/assets/c80d5a91-ccd9-478f-9666-ded2285e5d7a)

Reaction on archived messages:
![brave_F5WwW8BxjA](https://github.com/user-attachments/assets/25c33b21-b796-497f-96cd-d89b79ad4470)


In dark mode:
![image](https://github.com/zulip/zulip/assets/98596642/17f6ae40-7df0-4335-8f4f-ed0f73c477cb)
![image](https://github.com/zulip/zulip/assets/98596642/a1d122ea-5609-4e31-a49a-814a4afdb406)

![{BFDFAC3D-E36F-4236-8F94-5428A6024C97}](https://github.com/user-attachments/assets/9a759675-4a76-4ce3-af9a-359a979e1959)
![{60451150-357C-4655-A0AE-9803CF980B6D}](https://github.com/user-attachments/assets/6bdede3f-c1ff-49e0-8032-7a32b8b165fb)

![image](https://github.com/user-attachments/assets/703afda1-210e-4b72-ac7e-4a5c173830fa)


![{AC346B66-1502-4D94-8706-40DE01EE0E88}](https://github.com/user-attachments/assets/2058596d-76e3-4d72-9bbc-2fad12f8bd2e)
![{36762D06-7E4D-4AFD-9B2D-9246AA65CAE6}](https://github.com/user-attachments/assets/18fa2ea0-bb7b-41bb-bba8-9ba198efbf41)
![brave_8vfBjdlgpT](https://github.com/user-attachments/assets/35752d9c-186f-48a4-a4af-6a1390d7e67c)





<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sanchi-t commented on zerver/lib/display_recipient.py:
```diff
@@ -104,23 +105,45 @@ def get_tiny_stream_rows(
+                tag="stream-display-recipient",
+                stream_name=obj,
+                deactivated=False,
+            )
```
For some reason, the cache is only returning the stream name in one of the tests, specifically the one related to renaming a stream. I had to create an extractor function that defaults the deactivated value to false (We can also use Stream.objects.filter).

Initially, I wanted something like this: if obj has an incorrect datatype, then exclude those objects from cached_objects in cache.py. This would make the `generic_bulk_cached_fetch` to fetch values from the database. But I wasn't able to do that without making changes in `generic_bulk_cached_fetch`.

timabbott replied:
We probably need to debug this -- if everything we insert into the cache has a given type, then everything we pull out should have that type. (Especially in tests, where it should not be possible to have something stale in the cache from before this commit).

sanchi-t replied:
Yes, I will look into it.

timabbott replied:
What was the bug? It's helpful to always post that for posterity -- and it's possible that it reflects a related bug that needs to be fixed elsewhere.

sanchi-t replied:
I just needed to update the way the object was being inserted into the cache. Here:
```
cache_set(
        display_recipient_cache_key(recipient_id),
        {
            "tag": "stream-recipient-info",
            "stream_name": stream.name,
            "archived": stream.deactivated,
        },
    ) 
```

sanchi-t commented on web/src/stream_create.js:
```diff
@@ -290,7 +310,7 @@ function create_stream() {
                 // error text directly rather than turning it into
                 // "Error creating stream"?
-                stream_name_error.report_already_exists(stream_name);
+                stream_name_error.report_already_exists();
```
I don't believe this parameter was being passed, as report_already_exists had no parameters.

timabbott replied:
Indeed, the best way to handle this is to put this change in its own mini prep commit at the start of the branch with the explanation you had in this comment there in the commit message; I can then just merge that commit the next time I review this.

sanchi-t replied:
Fixed it.

timabbott replied:
Merged this one as 7eb34529e6fe11d483a74a91ca5fc2fdc5967914

timabbott commented on zerver/migrations/0497_remove_prefix_from_archived_streams.py:
```diff
@@ -0,0 +1,40 @@
+        prefix = hashed_stream_id + "!DEACTIVATED:"
+        prefix_length = len(prefix)
+        old_name = archived_stream.name
+        if old_name.startswith(prefix):
```
We should extend this migration to handful streams deactivated before 1b6f68bb59dc85125ff21fbd3f98a88f39c0730d.

sanchi-t replied:
Done.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -222,7 +208,8 @@ def deactivated_streams_by_old_name(realm: Realm, stream_name: str) -> QuerySet[
         # they provided:
-        Q(name__regex=rf"^{fixed_length_prefix}{truncated_name}")
+        Q(name=stream_name)
+        | Q(name__regex=rf"^{fixed_length_prefix}{truncated_name}")
```
Should this block of code be removed, given we don't expect to have remaining streams with `DEACTIVATED` in the name? Or is this to handle streams we can't properly rename in that migration?

This is part of  the `zerver/management/commands/unarchive_stream.py` management command, and not used for anything else.

sanchi-t replied:
> is this to handle streams we can't properly rename in that migration

Yes, since migration cannot guarantee the removal of the DEACTIVATED prefix from all archived streams, as many archived streams may retain their old names.

timabbott commented on zerver/models/streams.py:
```diff
@@ -180,6 +180,7 @@ def is_history_public_to_subscribers(self) -> bool:
     # * "realm" and "recipient" are not exposed to clients via the API.
     API_FIELDS = [
         "date_created",
+        "deactivated",
```
This is an API change that needs to be discussed and approved in the "API design" stream to coordinate with non-web clients. You can see the recent thread I started for a good example.

timabbott replied:
Started https://chat.zulip.org/#narrow/stream/378-api-design/topic/archived.20streams/near/1720667 for this.

timabbott commented on zerver/lib/types.py:
```diff
@@ -56,7 +56,13 @@ class UserDisplayRecipient(TypedDict):
+class StreamDisplayRecipient(TypedDict):
+    tag: Literal["stream-display-recipient"]
+    stream_name: str
+    deactivated: bool
```
What constraint requires us to add `deactivated` to display recipient objects? It would be kinda nice to avoid doing so, as we are generally working towards eliminating them in favor of just having the client use its own knowledge of the stream... though we'll need them in compatibility code for a long time, so it's not a near-term change.

sanchi-t replied:
Earlier I wrote something like:
`obj["deactivated"] = Stream.objects.get(id=recipient.type_id).deactivated)` in `zerver/lib/message.py`but it led to an extra query to the database. I will explore alternative methods to achieve this without fetching recipient objects for display.

sanchi-t replied:
Created a distinct type to distinguish it from the display recipient object, ensuring that eliminating display recipient objects won't significantly impact it. I've included fetching the `deactivated` status in the same query as obtaining the stream name for the display recipient to avoid an extra query. If this doesn't pose an issue, a separate function could be devised to retrieve stream status for stream messages. This approach would be a bit cleaner but comes with the tradeoff of increasing the cost.

timabbott commented on web/src/stream_settings_data.ts:
```diff
@@ -50,7 +50,7 @@ function get_subs_for_settings(subs: StreamSubscription[]): SettingsSubscription
     // (in bulk).  If that plan appears to have been aborted, feel free to
     // inline this.
-    return subs.map((sub) => get_sub_for_settings(sub));
+    return subs.filter((sub) => !sub.deactivated).map((sub) => get_sub_for_settings(sub));
```
This line seems out of place in the commit it's part of -- should it be an earlier independent commit for making deactivated streams that we're adding to the metadata not immediately appear in the settings UI?

sanchi-t replied:
Fixed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -19490,6 +19524,10 @@ components:
+        deactivated:
+          type: boolean
+          description: |
+            Only present for stream messages; indicating whether the stream is deactivated.
```
So these API documentation changes should go in the same commit where you added it to the actual API.

Additionally, they should have `**Changes**` entries with an API feature level bump; `git log api_docs/` should find you lots of handy examples. See also https://zulip.readthedocs.io/en/latest/documentation/api.html and the new application feature tutorial for examples.

sanchi-t replied:
Done.

timabbott commented:
@sanchi-t thanks for working on this! It looks like a promising start; posted a bunch of comments.

timabbott commented:
Thanks for all the effort on this. See https://chat.zulip.org/#narrow/stream/378-api-design/topic/archived.20streams/near/1720667 for a set of decisions on the API -- can you adjust the PR to implement those decisions? 

`git grep client_capabilities ` and `git grep bulk_message_deletion` should provide some helpful examples.

sanchi-t commented:
Yes. I'll soon update the PR.

sanchi-t commented:
From CZO discussion:

> In [/api/get-subscriptions](https://chat.zulip.org/api/get-subscriptions), no special action may be required -- you can't be still subscribed to an archived stream. We probably don't need an archived property on those objects, unless we want to enforce the invariant that subscription objects always extend the stream object?

So do we want to enforce the invariant that subscription objects always extend the stream object? Or `archived` property should not be included?

If we want not to include an `archived` property on get-subscriptions objects, we need to modify the type of subscriptions. Since both subscribed and unsubscribed streams share the same type (`SubscriptionStreamDict`), a straightforward solution is to make the property `archived: bool` optional in the type. This way, we can have the `archived` property specifically for unsubscribed streams.

timabbott commented:
> So do we want to enforce the invariant that subscription objects always extend the stream object? Or archived property should not be included?

For decisions about how the API should work like this, let's keep the discussion on the chat.zulip.org threads in "api design" -- that's where you'll get the right audience to make decisions.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,16 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 240**
+
+* [`GET /messages`](/api/get-message), [`GET /messages/{message_id}`](/api/get-message), `POST /zulip-outgoing-webhook`: Added a new field `archived` to messages, allowing client to mark messages from deactivated streams as archived.
```
These lines should be wrapped; I'm surprised `pretter` didn't do it for you, but your editor should have an appropriate feature. (E.g. `emacs` has meta-Q).

sanchi-t replied:
Yeah this should have not passed the lint test.
Anyway, I fixed it.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -897,10 +899,11 @@ def get_streams_for_user(
+    query = Stream.objects.filter(realm=user_profile.realm)
 
     if include_all_active:
+        query = query.filter(deactivated=False)
```
Hmm, this logic seems off for the case that one passed `include_all_active` and `include_archived`; I feel like that combination has to indicate they want all of them.

I think the logic we want is instead just for deactivated streams to be filtered, unless `include_archived` is passed, in which case they aren't -- i.e. a meaning of "don't exclude archived" rather than working like the other options. 

It's worth flagging this detail in the API design discussion though.

sanchi-t replied:
Yeah I improved the logic.

timabbott commented on zerver/lib/subscription_info.py:
```diff
@@ -57,6 +57,7 @@ def get_next_color() -> str:
 
         # Add versions of the Subscription fields based on a simulated
         # new user subscription set.
+        archived = False
```
This seems wrong -- `archived` is not a personal user setting -- it should just read this one out of `stream.deactivated`.

sanchi-t replied:
Yes you are right about that. That line should come under stream fields and not user subscription set.
I directly assigned `false` instead of getting value from `stream.deactivated` as web-public streams will never be `archived` but yeah maybe hardcoding here is not the right thing to do.

timabbott commented on zerver/lib/subscription_info.py:
```diff
@@ -428,6 +435,9 @@ def has_metadata_access_to_previously_subscribed_stream(
 
+    if stream_dict["archived"]:
+        return True
+
```
What's the reasoning behind this bit of permissions logic? I think we want this check after the `invite_only` check -- you shouldn't get extra access to a private stream you're not subscribed to just because it was archived after you were removed.

sanchi-t replied:
Well the problem lies in deactivating stream logic:
```
stream.deactivated = True
    stream.invite_only = True
    stream.save(update_fields=["deactivated", "invite_only"])
```
This change makes every archived stream invite-only, and only realm_admin will be able to view archived streams. We should maybe fix the logic for deactivating streams.

timabbott replied:
Yeah, I think once we integrate this PR, we'll want to do a follow-up PR to making archiving a stream not adjust its permissions. I had been imagining keeping that as a separate follow-up change as it's much more user-facing and will require more documentation changes, but it might be better to have it just be a final commit in this PR, since merging them the same week might be nice. 

I still think that means we drop this block of code.

sanchi-t replied:
Well I made some changes in archiving logic before your comment. I can undo those changes though since it will be better to have all that be done in one single PR. Just let me know if I should undo those changes.

timabbott commented on zerver/lib/subscription_info.py:
```diff
@@ -441,9 +451,10 @@ def has_metadata_access_to_previously_subscribed_stream(
 def gather_subscriptions_helper(
     user_profile: UserProfile,
     include_subscribers: bool = True,
+    archived_streams: bool = False,
```
I think `include_archived_streams` would be a clearer name for this parameter.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -154,6 +156,7 @@ def to_dict(self) -> Dict[str, Any]:
             pronouns_field_type_supported=self.pronouns_field_type_supported,
             linkifier_url_template=self.linkifier_url_template,
             user_list_incomplete=self.user_list_incomplete,
+            archived_streams=self.archived_streams,
```
I don't see this client capability doing anything in the events code. 

I think we'd [discussed](https://chat.zulip.org/#narrow/stream/378-api-design/topic/archived.20streams/near/1720672) that it should do this:

> and the events system pretend `archived: true` streams don't exist like the existing API does

But I'm not sure that's actually implementable easily -- the events code won't know if the stream is archived without a moderate amount of surgery. I'll raise that in the API design conversation.

sanchi-t replied:
I have removed the code related to it for now. I will update the PR with necessary changes following whatever is concluded from the discussion.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -19088,6 +19136,14 @@ components:
+            true if the stream is deactivated, false otherwise.
+
+            **Changes**: Changed in Zulip 9.0 (feature level 240).
+            Previously, subscriptions only included active streams.
```
So this description is a bit weird -- am I correct that this value is always `false`?

sanchi-t replied:
> am I correct that this value is always false, because we remove all subscribers as part of archiving a stream?

You are correct about that but the problem is that `unsubscribed` use the same schema as `Subscriptions`. So its not necessary that the value will always be false.
```
unsubscribed:
                        type: array
                        items:
                          $ref: "#/components/schemas/Subscriptions"
                        description: |
                          Present if `subscription` is present in `fetch_event_types`.
```

timabbott replied:
OK I think that's a good reason for having the field be always present on subscribed objects, thanks for tracking that down!

timabbott commented on zerver/actions/streams.py:
```diff
@@ -254,7 +254,14 @@ def do_unarchive_stream(
+            "stream_name": stream.name,
+            "archived": stream.deactivated,
+        },
+    )
```
I still don't understand this change -- why do we need to mark stream messages as "archived" at all? I don't think it'll be possible to send a message to an archived stream, and clients should be able to look up via the `stream` data structures  whether a given stream that they are processing a message from is archived.

I suspect we just don't need this commit, but maybe I'm missing something.

sanchi-t replied:
Ah I see. I misunderstood the situation and now  have made the necessary changes. 
The `archived` property has been removed from messages, and the client now checks the stream's archived status to display the "(archived)" postfix for deactivated streams.

timabbott commented on web/templates/stream_settings/stream_creation_form.hbs:
```diff
@@ -12,6 +12,7 @@
                     <input type="text" name="stream_name" id="create_stream_name" class="settings_text_input"
                       placeholder="{{t 'Stream name' }}" value="" autocomplete="off" maxlength="{{ max_stream_name_length }}" />
                     <div id="stream_name_error" class="stream_creation_error"></div>
+                    <a id="archived_stream_rename"></a>
```
I don't see a backend change adding support for editing these -- is it correct then that the API in `main` allows administrators to rename archived streams, and it just wasn't an exposed option in the UI?  

At least my read of `access_stream_for_delete_or_update` is that this is the case, but worth confirming.

sanchi-t replied:
The same API used for renaming streams in the `streams` menu is also utilized for renaming an archived stream. This specific function is responsible for granting access to rename a stream, and only an admin is allowed to make such a change.
```
def check_stream_access_for_delete_or_update(
        user_profile: UserProfile, stream: Stream, sub: Optional[Subscription] = None
    ) -> None:
        error = _("Invalid stream ID")
        if stream.realm_id != user_profile.realm_id:
            raise JsonableError(error)
   
        if user_profile.is_realm_admin:
            return
   
       if sub is None and stream.invite_only:
           raise JsonableError(error)
  
       raise OrganizationAdministratorRequiredError
```

timabbott commented:
@sanchi-t please do a top-level comment with a changelog and any outstanding conversations when you've addressed this last batch of feedback fully; see @prakhar1144's PRs for some good examples of how to communicate really clearly about updates.

(And maybe bump any API design conversations that are still not decided).

sanchi-t commented:
@timabbott this pull request isn't quite ready yet. I should have marked it as a draft. I'll notify you in the stream as soon as it is ready for review.

sanchi-t commented on web/src/message_list.js:
```diff
@@ -344,7 +344,7 @@ export class MessageList {
         const can_toggle_subscription =
             sub !== undefined && stream_data.can_toggle_subscription(sub);
-        if (sub === undefined) {
+        if (sub.archived || sub === undefined) {
```
With deactivated streams now being returned to the client, could sub be `undefined`? If so, should we assume it's deactivated?

Also we should rename `deactivated` to `archived` maybe but that should be done in a follow up PR ig.

timabbott replied:
I think it is still possible for `sub` to be undefined, but only in the preserve of bugs. Probably a follow-up project for sure.

And yes, we should plan to do follow-up passes renaming `deactivated to `archived`, both here and in the server implementation -- but each should be its own project, since the server one will require a database migration (and thus a lot of work), while the web app rename should be an easy change.

sanchi-t commented:
Updates in the recent push:

- Addressed the comments.
- Fixed the stream archiving logic which made every archived stream `invite_only: true`.
- Removed the use of `display_recipient` for sending the `archived` status for stream messages.
- Removed events system changes regarding archived streams.

Key conversation points:
- `archived` property should exist for subscription objects as `unsubscribed` object will potentially contain streams that are archived.

CZO discussion: https://chat.zulip.org/#narrow/stream/378-api-design/topic/archived.20streams/near/1720667

sanchi-t commented on web/tests/dispatch_subs.test.js:
```diff
@@ -236,6 +240,8 @@ test("stream delete (special streams)", ({override}) => {
 
     for (const stream of event.streams) {
+        stream.archived = true;
+        stream.subscribed = true;
```
I don't completely understand this whole testcase.
Please be a bit more thorough during review.

sanchi-t replied:
Currently, deactivating/archiving a stream involves updating the `archived: true` and `subscribed: false` in the stream objects. 

In the given testcase, `test("stream delete (normal)")` precedes the modification of values(`archived` and `subscribed`) in `event_fixtures.stream__delete`.

Consequently, for `test("stream delete (special streams)")`, where `remove_sidebar_row` is expected to be called, the streams in `event_fixtures.stream__delete` need to be marked as `archived: false` and `subscribed: true`, as remove_sidebar_row is only triggered for `subscribed: true` streams.

Here:
```
if (was_subscribed) {
   stream_list.remove_sidebar_row(stream.stream_id);
}
```

But I see how I have wrongly written `archived: true`. I will be more careful.

sanchi-t commented on web/src/server_events_dispatch.js:
```diff
@@ -546,20 +547,22 @@ export function dispatch_normal_event(event) {
-                        stream_data.delete_sub(stream.stream_id);
                         stream_settings_ui.remove_stream(stream.stream_id);
+                        message_view_header.maybe_rerender_title_area_for_stream(stream);
+                        message_live_update.rerender_messages_view();
                         if (was_subscribed) {
                             stream_list.remove_sidebar_row(stream.stream_id);
                             if (stream.stream_id === compose_state.selected_recipient_id) {
                                 compose_state.set_selected_recipient_id("");
                                 compose_recipient.on_compose_select_recipient_update();
                             }
                         }
-                        settings_streams.update_default_streams_table();
                         stream_data.remove_default_stream(stream.stream_id);
+                        settings_streams.update_default_streams_table();
```
I think the logical order this function is wrong.

`delete_sub` must be called before `update_trailing_bookend`; otherwise, an incorrect trailing end will be presented.

If a stream is opened and subsequently deactivated, the following trailing end used to be visible:
![image](https://github.com/zulip/zulip/assets/98596642/9b2f4a4c-f505-4319-bc6b-76b9107e753c)

That's why I was surprised when I discovered the existence of a specific trailing end for deactivated streams. Despite numerous interactions with deactivating and deactivated streams, I never saw it, partly due to this incorrect order.

Likewise `remove_default_stream` should be called before `update_default_streams_table` since `update_default_streams_tables` invokes `build_default_stream_table`  which uses `default_stream_ids` before it is updated.

timabbott replied:
Good finds!

sanchi-t replied:
Done.

sanchi-t commented on zerver/lib/events.py:
```diff
@@ -1089,7 +1092,7 @@ def _draft_update_action(i: int) -> None:
                 stream
                 for stream in state["unsubscribed"]
-                if stream["stream_id"] not in deleted_stream_ids
+                if has_metadata_access_to_previously_subscribed_stream(user_profile, stream)
```
Is this fine?

timabbott replied:
Don't you also need to update the `archived` flag for streams that remain in this list, for example?

You should write a `test_events.py` test that verifies the case where an administrator unsubscribes from a stream and then archives it.

sanchi-t replied:
Fixed.

sanchi-t replied:
We no longer need to update the `archived` flag and `subscribers` list for streams that remain in this list as #30315 has been stacked into this PR and #30315 was responsible for fixing the incorrect stream logic where channels are incorrectly marked as `invite_only: true` and removed all the subscribers from channel when archiving.

sanchi-t commented:
Updates since last review:

- Addressed the comments.
- Addressed the CZO comments.
- Fixed the stream archiving logic which made every archived stream `invite_only: true`.
- Removed the use of `display_recipient` for sending the archived status for stream messages.
- Removed events system changes regarding archived streams.
- Fixed handling of deactivating stream in client.
- The stream name is now updated with the `(Archived)` suffix immediately after deactivating it.

**NOTE:** I have excluded the event system changes (modifications made in Tornado) since passing the 'archived' parameter to Tornado code has no effect. I may not fully comprehend the meaning behind:

> So it's harmless to send those for archived streams even when those clients haven't set the capability.

We referring to the event system here right? However, I don't see the purpose of sending the 'archived' flag to the event system.


Key conversation points:

 - `archived` property should exist for subscription objects as `unsubscribed` object will potentially contain streams that are archived.
CZO discussion: https://chat.zulip.org/#narrow/stream/378-api-design/topic/archived.20streams/near/1720667

timabbott commented on zerver/migrations/0500_remove_prefix_from_archived_streams.py:
```diff
@@ -0,0 +1,51 @@
+            archived_stream.save(update_fields=["name"])
+
+
+class Migration(migrations.Migration):
```
This migration should be set with `atomic=False`, so that if there's many thousands of streams, it doesn't do them all in a single transaction.

sanchi-t replied:
Fixed.

timabbott commented on zerver/migrations/0500_remove_prefix_from_archived_streams.py:
```diff
@@ -0,0 +1,51 @@
+            continue
+
+        # Check if there's an active stream or another archived stream with the new name
+        if not Stream.objects.filter(name=new_name).exists():
```
This filter needs to be limited to the target realm.

sanchi-t replied:
Fixed.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -145,21 +143,7 @@ def do_deactivate_stream(stream: Stream, *, acting_user: Optional[UserProfile])
-    new_name = (hashed_stream_id + "!DEACTIVATED:" + old_name)[: Stream.MAX_NAME_LENGTH]
-
-    stream.name = new_name[: Stream.MAX_NAME_LENGTH]
-    stream.save(update_fields=["name", "deactivated", "invite_only"])
```
The commit message needs to be very clear that this also no longer marks them as `invite_only`.

sanchi-t replied:
Okay so I undo my changes. I think a second PR for this change will be better.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -658,7 +658,7 @@ def filter_stream_authorization(
 
         # Members and administrators are authorized for public streams
-        if not stream.invite_only and not user_profile.is_guest:
+        if not stream.invite_only and not user_profile.is_guest and not stream.deactivated:
```
Why is this security/authorization change in this commit? It's very concerning to see this squashed in with a commit about name changes, because that's a common way to accidentally introduce security bugs.

This reinforces my opinion in https://github.com/zulip/zulip/pull/28478/commits/0601fc653b6d87224869d95ad54c23d5ab3fa163#r1500104925 that you should remove the `invite_only` changes from this commit.

sanchi-t replied:
I agree with you. We should leave it for a next PR.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,21 @@ format used by the Zulip server that they are interacting with.
+* [`POST /register`](/api/register-queue): The new [client capability]
+  (/api/register-queue#parameter-client_capabilities) `archived_streams`
+  allows the client to decide whether to include archived streams in the
+  result.
```
The commit message for this commit still incorrectly talks about a `deactivated` field, when it's called `archived`.

sanchi-t replied:
Done.

timabbott commented on zerver/lib/events.py:
```diff
@@ -1636,6 +1641,7 @@ def do_events_register(
         pronouns_field_type_supported=pronouns_field_type_supported,
         linkifier_url_template=linkifier_url_template,
         user_list_incomplete=user_list_incomplete,
+        archived_streams=archived_streams,
```
Doesn't this need to be passed to `request_event_queue` too?

sanchi-t replied:
I mean it is harmless to pass `archived_stream` parameter but I don't see the point. Like what will that do?
Anyway I updated the code according to you.

timabbott commented on zerver/lib/home.py:
```diff
@@ -153,6 +153,7 @@ def build_page_params_for_home_page_load(
         "user_settings_object": True,
         "linkifier_url_template": True,
         "user_list_incomplete": True,
+        "archived_streams": True,
```
I think in this first commit, where the web app doesn't actually support seeing archived streams, this should probably be `False`? That way, this commit works.

sanchi-t replied:
Yeah makes sense. Fixed it.

timabbott commented on zerver/lib/subscription_info.py:
```diff
@@ -111,6 +113,7 @@ def build_unsubscribed_sub_from_stream_dict(
     raw_stream_dict = RawStreamDict(
         can_remove_subscribers_group_id=stream_dict["can_remove_subscribers_group"],
         date_created=timestamp_to_datetime(stream_dict["date_created"]),
+        deactivated=Stream.objects.get(id=stream_dict["stream_id"]).deactivated,
```
Yikes! This will do a database query for each stream processed this way. Isn't the value already in the `stream_dict`? This seems like a potentially significant performance bug.

sanchi-t replied:
That's right. Fixed it.

timabbott commented on zerver/lib/subscription_info.py:
```diff
@@ -441,9 +448,10 @@ def has_metadata_access_to_previously_subscribed_stream(
 ) -> SubscriptionInfo:
     realm = user_profile.realm
-    all_streams = get_active_streams(realm).values(
+    all_streams = get_all_streams(realm, include_archived_streams).values(
```
Spell this as `get_all_streams(realm, include_archived_streams=include_archived_streams)`: passing via kwargs makes it obvious to someone reading this code what that second parameter does.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11935,6 +11944,15 @@ paths:
+                    - `archived_streams`: Boolean for whether the client's
+                      handler for the `stream` event type want to get results
+                      including archived streams. Otherwise, the server will
+                      send `stream` events only for non-archived streams.
```
This should use the same writing style as the others.

"Boolean for whether the client supports processing archived streams in the `stream` and `subscription` event types. If `false`, the server will not include in the `register` response or in events related to archived streams."

sanchi-t replied:
Fixed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -16974,6 +16993,14 @@ paths:
+        - name: exclude_archived
+          in: query
+          description: |
+            Excludes archived streams in realm.
```
"Whether to exclude archived streams from the results."

The writing of the existing options isn't great on this page, but the "in realm" addition in particular is unnecessary/confusing, since no Zulip API lets you access data outside your realm.

sanchi-t replied:
Fixed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -18578,6 +18614,14 @@ components:
+          type: boolean
+          description: |
+            A boolean indicating whether the stream is deactivated;
+            true if the stream is deactivated, false otherwise.
```
I think this would be ideal:

> A boolean indicating whether a stream is [archived](/help/archive-a-stream).

sanchi-t replied:
Done.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -18578,6 +18614,14 @@ components:
+            true if the stream is deactivated, false otherwise.
+
+            **Changes**: Changed in Zulip 9.0 (feature level 242).
+            Previously, only active streams were sent.
```
I suggest the following:

> New in Zulip 9.0 (feature level 242). Previously, this API never returned archived streams.

p.s. you could have guessed this if you looked at out conventions for other `Changes` entries.

sanchi-t replied:
Done.

timabbott commented on web/src/stream_settings_data.ts:
```diff
@@ -50,7 +50,7 @@ function get_subs_for_settings(subs: StreamSubscription[]): SettingsSubscription
     // (in bulk).  If that plan appears to have been aborted, feel free to
     // inline this.
-    return subs.map((sub) => get_sub_for_settings(sub));
+    return subs.filter((sub) => !sub.archived).map((sub) => get_sub_for_settings(sub));
```
Why is this change the right place to filter? I would expect to want to do the filtering in higher-level functions closer to where we're actually using them.

sanchi-t replied:
I think it is the right place to filter. Both functions `get_streams_for_settings_page` and `render_left_panel_superset` depend on `get_subs_for_settings` to get stream related data.
We can  filter streams separately within each function but I don't think its a very good approach.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -546,20 +546,22 @@ export function dispatch_normal_event(event) {
-                        stream_data.delete_sub(stream.stream_id);
                         stream_settings_ui.remove_stream(stream.stream_id);
+                        message_view_header.maybe_rerender_title_area_for_stream(stream);
+                        message_live_update.rerender_messages_view();
```
Can this line be pullet out of the loop? It's a very expensive operation.

sanchi-t replied:
Done.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -546,20 +546,22 @@ export function dispatch_normal_event(event) {
                             }
                         }
-                        settings_streams.update_default_streams_table();
                         stream_data.remove_default_stream(stream.stream_id);
```
Presumably this data update could be at the very start?

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -546,20 +546,22 @@ export function dispatch_normal_event(event) {
                         }
-                        stream_data.delete_sub(stream.stream_id);
                         stream_settings_ui.remove_stream(stream.stream_id);
+                        message_view_header.maybe_rerender_title_area_for_stream(stream);
```
I think this line should be added 2 commits later, when you make the title area rendering depend on whether the stream is archived?

sanchi-t replied:
I think it will be better to have this whole commit later.

timabbott commented on web/src/message_list_view.js:
```diff
@@ -219,6 +219,7 @@ function populate_group_from_message_container(group, message_container) {
         group.topic = message_container.msg.topic;
         group.match_topic = util.get_match_topic(message_container.msg);
         group.stream_url = message_container.stream_url;
+        group.archived = stream_data.is_archived(message_container.msg.stream_id);
```
Let's call this `is_stream_archived` -- we should scope it clearly as not a "archived message".

sanchi-t replied:
Done.

timabbott commented on web/src/stream_data.ts:
```diff
@@ -646,6 +646,11 @@ export function get_stream_privacy_policy(stream_id: number): string {
+export function is_archived(stream_id: number): boolean {
+    const sub = sub_store.get(stream_id);
+    return sub ? sub.archived : false;
+}
```
This file expects 100% test coverage because it only contains easily tested logic, you should just include a test for the new function rather than adding the file to an exclude list.

sanchi-t replied:
Fixed.

timabbott commented on web/styles/app_variables.css:
```diff
@@ -231,6 +231,7 @@
     /* Text colors */
     --color-text-default: hsl(0deg 0% 20%);
     --color-text-message-default: hsl(0deg 0% 15%);
+    --color-text-message-header-archived: hsl(0deg 0% 50%);
```
How did you pick this color, and do we need a different color for the dark theme?

timabbott commented on web/src/stream_create.js:
```diff
@@ -58,15 +61,16 @@ class StreamSubscriptionError {
-            $t({defaultMessage: "A stream with this name already exists"}),
-        );
+    report_already_exists(error) {
+        const errorMessage =
```
Use `error_message` -- we don't use camelCase variable names.

sanchi-t replied:
Yeah I was aware of that but looking at `defaultMessage` I thought in this specific instance, I needed to use camelCase.
Fixed though.

timabbott commented on web/src/stream_create.js:
```diff
@@ -78,6 +82,12 @@ class StreamNameError {
 
+    rename_archived_stream(stream_name) {
+        $("#archived_stream_rename").text($t({defaultMessage: "Rename archived stream"}));
+        $("#archived_stream_rename").attr("data-stream-name", stream_name);
```
Can we use a `data-stream-id` instead? I guess this may be a case where we really do need the name; we just try to avoid working with mutable stream names where possible.

sanchi-t replied:
Yes we can.
I updated the function.

timabbott commented on web/src/stream_create.js:
```diff
@@ -86,8 +96,18 @@ class StreamNameError {
+                }
+            } else {
+                this.report_already_exists();
+            }
```
We probably should check that you have permission to administer the stream in question?

sanchi-t replied:
I am not sure I fully understand.
`can_change_name_description` makes sure admin has the access to rename the stream.
And if someone with the permission to create streams attempts to create a stream with the same name as an existing one, they will receive an error message stating:

> A stream with this name already exists

Why is it necessary to check permissions before showing this error? If a user has the ability to create a stream and is attempting to create one with a name already in use, they should be informed that the name is already taken.

timabbott commented on tools/test-js-with-node:
```diff
@@ -241,6 +242,7 @@ EXEMPT_FILES = make_set(
         "web/src/stream_settings_containers.ts",
         "web/src/stream_settings_ui.js",
         "web/src/stream_ui_updates.js",
+        "web/src/sub_store.ts",
```
So I mentioned this for `stream_data.ts`, but this should also be address instead by changing the tests.

More generally, this commit shouldn't exist -- each individual commit should pass all tests, and the changes to tests required

Check out the [Zulip commit message](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for more details -- but one reason this practice is important is that it makes it possible to merge or revert individual commits.

sanchi-t replied:
Okay, I will remember it for future pull requests.

timabbott commented on web/tests/stream_data.test.js:
```diff
@@ -496,15 +496,17 @@ test("delete_sub", () => {
-    assert.ok(!sub_store.get(canada.stream_id));
+    assert.ok(stream_data.get_sub("Canada"));
+    assert.ok(sub_store.get(canada.stream_id));
+    assert.equal(stream_data.num_subscribed_subs(), num_subscribed_subs - 1);
```
So just add a couple calls here to check whether the Canada stream is archived before and after deactivating it, and this test will have 100% coverage of `stream_data.ts` again.

sanchi-t replied:
Done.

timabbott commented on zerver/lib/subscription_info.py:
```diff
@@ -438,6 +438,9 @@ def has_metadata_access_to_previously_subscribed_stream(
+    if stream_dict["archived"]:
+        return True
+
     return True
```
Can you explain what this test does, given that you `return True` just after anyway?

sanchi-t replied:
Yes you are right. Fixed that.

timabbott commented:
Thanks for all the hard work on this @sanchi-t! I posted what I think is a complete review for the remaining changes.

sanchi-t commented on zerver/migrations/0500_remove_prefix_from_archived_streams.py:
```diff
@@ -0,0 +1,51 @@
+
+    for archived_stream in archived_streams:
+        old_prefix = "!DEACTIVATED:"
+        streamID = str(archived_stream.id)
```
I don't think so, looking at the stream model:
```
API_FIELDS = [
        ...
        "id",
        ...
]
```
Additionally, you'll encounter an error if you attempt to use `stream_id` instead of `id`:
`AttributeError: 'Stream' object has no attribute 'stream_id'`.

I believe `stream_id`  is specific to APIs, as indicated by this comment:
`# * "id" is represented as "stream_id" in most API interfaces.`



On a side note:
This comment from `zerver/models/streams.py` is not right anymore is it?
`# * "deactivated" streams are filtered from the API entirely.`

alya commented:
Thanks! A bunch of UI tweaks here:

1. For the stream header bar label, it would be nice to use consistent styling with guest users:
<img width="268" alt="Screenshot 2024-03-07 at 1 50 10 PM" src="https://github.com/zulip/zulip/assets/2090066/5137373f-c35c-4231-819b-bea896173c5b">

I actually like the dimming, so we can try making it "_(archived)_"  but keeping that part of the styling for now, and we can get more feedback on that in #design.

alya commented:
2) For the edit modal, can we change the title to something like this? We need some clear indication that you're editing the archived stream, not the stream you're trying to create or something.

> Edit #Denmark (archived stream)

3) I managed to get into this state.

<img width="525" alt="Screenshot 2024-03-07 at 2 01 11 PM" src="https://github.com/zulip/zulip/assets/2090066/1ee2ebee-5105-483b-a064-e5f4130f0901">

This was after a weird flow, where I was not on this PR, I archived #Denmark, made a new #Denmark, archived that one, and then got the PR running and tried to rename it.

So maybe this is just a transition edge case to watch out for?

alya commented:
4) After I renamed the archived stream, the error message ("An archived stream with this name already exists.") and renaming link did not disappear like they should.

alya commented:
5. Archived streams appear with the private stream icon , which seems wrong?

<img width="390" alt="Screenshot 2024-03-07 at 2 09 54 PM" src="https://github.com/zulip/zulip/assets/2090066/9f3b7eb9-8818-41c4-933e-78869ac85f6b">


6. The archived stream isn't displayed correctly to a non-admin user:

<img width="365" alt="Screenshot 2024-03-07 at 2 08 32 PM" src="https://github.com/zulip/zulip/assets/2090066/3fa7a7ff-aa56-4192-b18c-0eb98f9fe2fb">

Clicking on a message produces an error:

<img width="959" alt="Screenshot 2024-03-07 at 2 11 38 PM" src="https://github.com/zulip/zulip/assets/2090066/8f91f0ab-db5f-4ee4-998d-ed0667908d32">


And clicking on the stream or topic looks like this:

<img width="915" alt="Screenshot 2024-03-07 at 2 11 46 PM" src="https://github.com/zulip/zulip/assets/2090066/241ae965-91c0-4bcd-a7e6-0bcd3d26f911">

alya commented:
I'll probably have more feedback, but I think this is enough to work through for now.

sanchi-t commented:
@alya  I have responded to your comments, however I need to details to figure out a few things.



1) > I actually like the dimming
 By dimming you mean keeping the gray color for now right? I have updated the PR.

2) Done.

3) Okay so I don't fully understand how you managed to get into that state.

Before the PR, when deactivating a stream, its name was transformed into a hashed stream ID + "!DEACTIVATED" + stream.name. This allowed creating #Denmark multiple times without errors since each deactivation appended a unique prefix.

Now, after running this PR, the migration restores stream names to their original state. When running the migration, you'll notice one archived #Denmark stream with the name #Denmark only, while the other retains the prefix. Due to the constraint of unique stream names, the migration skips one stream.

Attempting to create another stream named #Denmark will trigger an error, indicating "An archived stream with this name already exists." The error provides an option to rename the archived stream.

Regarding the "Failed: Invalid parameter" error, I couldn't reproduce it. Can you provide more details on when it occurred? Ensure that you ran the "removing_prefix_from_archived_streams" migration after switching to this PR and reloaded.

The message in the original #Denmark will have that prefix because, as mentioned in issue #26661, the migration skips streams with the same name:

> Skip renaming if there is another stream in the organization that has the same name.

4) I'm unable to replicate this issue either. Could you clarify if you attempted to rename the #Denmark stream mentioned in the third point or a different one? The error message disappears for me, so additional details would help diagnose the problem accurately.
Here:
[rename.webm](https://github.com/zulip/zulip/assets/98596642/7ce2916f-033e-40b0-98d0-50dd8bb19898)


5) The streams appear with the private stream icon due to an issue in the stream deactivating logic, which incorrectly marks the stream as private. We have acknowledged this problem and agreed to address it in a subsequent PR. This issue was highlighted by Tim in his comment.

> Yeah, I think once we integrate this PR, we'll want to do a follow-up PR to making archiving a stream not adjust its permissions.


6) Since archived streams are currently marked as private, they will only be visible to admins. Non-admin users will still be able to view messages from archived streams in search results, just as they could before. However, they will not have access to view the stream topics themselves.

It's worth noting that the error you encountered when clicking on a message from an archived stream is not a result of this PR. This error was present prior to the implementation of this PR.

As for the "(archived)" suffix not being visible to non-admin users, this is due to the fact that non-admin users do not have access to archived streams. Consequently, the client cannot determine whether a stream is
```
export function is_stream_archived(stream_id: number): boolean {
    const sub = sub_store.get(stream_id);
    return sub ? sub.archived : false;
}
```

@timabbott Should I address the issue of getting an error (`adjusted_terms impossibly null`) when clicking on a message from an archived stream within this PR itself?

alya commented:
> Ensure that you ran the "removing_prefix_from_archived_streams" migration after switching to this PR and reloaded.

How would I do that?

alya commented:
> I'm unable to replicate this issue either. Could you clarify if you attempted to rename the #Denmark stream mentioned in the third point or a different one? The error message disappears for me, so additional details would help diagnose the problem accurately.

Ah, sorry this was unclear. This happens for me in general, but not all the time, so I'm not sure what the conditions are.

alya commented:
The experience for editing, moving or deleting a message is currently confusing; messages should appear as not editable/movable/deletable in the UI. Let's file an issue if it makes sense to change this as a separate follow-up PR.

sanchi-t commented:
> How would I do that?

You can simply run the command `./manage.py migrate`. This command will apply all the unapplied migrations.

And if for some reason you can not apply the `0501_remove_prefix_from_archived_streams.py` migration, you need to reverse the migration first then apply it again.

You can do this using this command:
`./manage.py migrate zerver 0500`

Where 0500 is the migration number before the migration you want to apply, which in this case is 0501.

sanchi-t commented:
> Ah, sorry this was unclear. This happens for me in general, but not all the time, so I'm not sure what the conditions are.

Ah, that's alright I will keep testing it hopefully I will be able to reproduce it.

sanchi-t commented on web/src/stream_settings_ui.js:
```diff
@@ -123,6 +123,9 @@ export function update_stream_name(sub, new_name) {
 
     // Update navbar if needed
     message_view_header.maybe_rerender_title_area_for_stream(sub);
+
+    // Update the create stream error if needed
+    stream_create.maybe_update_error_message();
```
I think this will help with the bug Alya was facing. 

> After I renamed the archived stream, the error message ("An archived stream with this name already exists.") and renaming link did not disappear like they should.


Although I was never able to reproduce the bug, I think the bug could be fixed by updating the error message in server events.

sanchi-t commented:
Updates since last review:

- Changed `archived` field to `is_archived`.
- Fixed the bug where the error message ("An archived stream with this name already exists.") and renaming link were not disappearing after renaming in create stream.

sanchi-t commented:
Here is what PR implements:

- Changes made to the archiving logic of streams; no renaming or prefix addition when a stream is archived.
- Archived streams are now sent back to the client; previously, they were excluded.
- Archived streams are now included in the unsubscribed streams list.
- Archived streams are filtered out and not displayed in settings menu.
- APIs '/streams/{stream_id}' and '/events' now include an `is_archived` property for streams and subscriptions.
- For '/streams', an `exclude_archived` parameter is added.
- For '/register', a client capability parameter `archived_streams` is added.
- Add a new test for deactivating a stream.
- Show an '(archived)' label next to archived stream names in the client, similar to the '(guests)' label.
- Allow users with permissions to rename an archived stream when creating a new stream with the same name.
- NOTE: There's a problem with the stream archiving process that makes archived streams `invite_only: true`, making them invisible to non-admin users. This will be fixed in a follow up PR.

timabbott commented on web/src/stream_create.js:
```diff
@@ -520,3 +546,51 @@ export function set_up_handlers() {
+        e.stopPropagation();
+        const stream_id = Number.parseInt($("#archived_stream_rename").attr("data-stream-id"), 10);
+        const stream = stream_data.get_sub_by_id(stream_id);
+        const stream_name = stream.name + " (archived stream)";
```
You can't append diff like this. This works:

``` diff
diff --git a/web/src/stream_create.js b/web/src/stream_create.js
index 5808d1bc5b..b5b38a4c8e 100644
--- a/web/src/stream_create.js
+++ b/web/src/stream_create.js
@@ -553,7 +553,6 @@ export function initialize() {
         e.stopPropagation();
         const stream_id = Number.parseInt($("#archived_stream_rename").attr("data-stream-id"), 10);
         const stream = stream_data.get_sub_by_id(stream_id);
-        const stream_name = stream.name + " (archived stream)";
         const template_data = {
             stream_name: stream.name,
             stream_description: stream.description,
@@ -562,7 +561,8 @@ export function initialize() {
         };
         const change_stream_info_modal = render_change_stream_info_modal(template_data);
         dialog_widget.launch({
-            html_heading: $t_html({defaultMessage: "Edit #{stream_name}"}, {stream_name}),
+            html_heading: $t_html({defaultMessage: "Edit #{stream_name} (<i>archived</i>)"},
+                                  {stream_name: stream.name}),
             html_body: change_stream_info_modal,
             id: "change_stream_info_modal",
             loading_spinner: true,
```

timabbott commented:
I pushed back here after doing the adjustments noted in https://chat.zulip.org/#narrow/stream/378-api-design/topic/archived.20streams/near/1946212 as well as squashing a few commits that were not coherently separated and including the commits in #30315 here so we can merge them together. I also added a basic help/ commit.

There's a test failing that I don't understand.

alya commented:
In the confirmation modal: "Archiving channel" -> "Archiving this channel".

alya commented:
Archiving a channel and pressing "back" a couple of times in the browser, I got:

![Screenshot 2024-09-23 at 13 52 58@2x](https://github.com/user-attachments/assets/3844e8f6-6051-4624-aef0-a8f307ec098f)

alya commented:
Message actions need to be handled more gracefully. Can we

1. Hide the actions you can't take from the message menu, message header bar, and anywhere else?
2. Show a more useful error if someone attempts to take the action (e.g., another user who hasn't refreshed the page)?

Some examples:
![Screenshot 2024-09-23 at 13 54 56@2x](https://github.com/user-attachments/assets/fa73ff1d-8159-4a2a-8d65-3abbc3398723)

![Screenshot 2024-09-23 at 13 55 05@2x](https://github.com/user-attachments/assets/a0915dbf-7119-422a-81d3-8518ce40a2dc)

![Screenshot 2024-09-23 at 13 58 34@2x](https://github.com/user-attachments/assets/ed661dca-7c16-4c59-b6a7-2a36e07324f6)

Emoji reaction:
![Screenshot 2024-09-23 at 13 55 33@2x](https://github.com/user-attachments/assets/085df6fd-7040-4602-ae39-ed558749fef6)

The "resolve topic" button in the message header bar simply vanishes when you click it.

alya commented:
Apologies if some of this was discussed above; I haven't had a chance to review this comment thread.

sanchi-t commented:
> In the confirmation modal: "Archiving channel" -> "Archiving this channel".

Done.

>Archiving a channel and pressing "back" a couple of times in the browser, I got:

Resolved.
